In [2]:
import pandas as pd
import json

In [4]:
pd.DataFrame

pandas.core.frame.DataFrame

In [5]:
with open ("chessbuds_messages.json") as c:
    chess = json.load(c) 

In [7]:
chess.keys()

dict_keys(['participants', 'messages', 'title', 'is_still_participant', 'thread_type', 'thread_path', 'magic_words', 'joinable_mode'])

In [8]:
name_df = pd.DataFrame(chess['participants'])

In [9]:
name_df

name
0           Scott Pence
1           Chad Larson
2          Joanna Rusch
3  Angela Babbitt Pence
4           David Silva
5           Aaron Rusch
6    Timothy Vanderpool

In [11]:
type(name_df)

pandas.core.frame.DataFrame

In [13]:
chess_df = pd.DataFrame(chess['messages'])

In [14]:
type(chess_df)

pandas.core.frame.DataFrame

In [16]:
len(chess_df)

223

In [19]:
type(chess['title'])

str

In [21]:
chess['title']

'Chess Buds'

In [22]:
len(chess['title'])

10

In [23]:
type(chess['joinable_mode'])

dict

In [24]:
len(chess['joinable_mode'])

2

In [25]:
chess['joinable_mode']

{'mode': 1, 'link': ''}

In [27]:
chess_df.head()

sender_name   timestamp_ms  \
0  Joanna Rusch  1666374933946   
1   Chad Larson  1666373448613   
2   Chad Larson  1666373216381   
3   Scott Pence  1666373164883   
4   Chad Larson  1666373111157   

                                             content  \
0  Maybe he just wants to ride the publicity for ...   
1  To be fair to Hans....no one wants to be assoc...   
2  He would have to prove he didn't cheat and tha...   
3  Yeah, no way. You over shoot and hope to get a...   
4  From what I see, I don't think he could win.  ...   

                                           reactions     type  is_unsent  \
0  [{'reaction': '👍', 'actor': 'Chad Larson'},...  Generic      False   
1  [{'reaction': '😆', 'actor': 'Scott Pence'},...  Generic      False   
2  [{'reaction': '👍', 'actor': 'Scott Pence'},...  Generic      False   
3  [{'reaction': '👍', 'actor': 'Chad Larson'},...  Generic      False   
4                                                NaN  Generic      False   

   is_taken_down                            bumped_message_metadata share  \
0          False  {'bumped_message': 'Maybe he just wants to rid...   NaN   
1          False  {'bumped_message': 'To be fair to Hans....no o...   NaN   
2          False  {'bumped_message': 'He would have to prove he ...   NaN   
3          False  {'bumped_message': 'Yeah, no way. You over sho...   NaN   
4          False  {'bumped_message': 'From what I see, I don't t...   NaN   

  photos gifs users  
0    NaN  NaN   NaN  
1    NaN  NaN   NaN  
2    NaN  NaN   NaN  
3    NaN  NaN   NaN  
4    NaN  NaN   NaN

In [28]:
num_reactions = [len(x) for x in chess_df['reactions']]

TypeError: object of type 'float' has no len()

Let me resolve the above issue first, to properly calcuate the lengths of each entry in the 'reactions' coloumn, the Nan entries (or holes) need to be replaced with empty strings or 0s first. Listed Pandas API reference section, the fillna method can be used to fulfill this purpose (url=https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html). 

In [29]:
chess_df['reactions'].fillna('', inplace=True)

In [30]:
num_reactions = [len(x) for x in chess_df['reactions']]

In [31]:
num_reactions

[2,
 2,
 2,
 2,
 0,
 2,
 2,
 0,
 4,
 4,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 4,
 0,
 0,
 0,
 4,
 0,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 2,
 0,
 0,
 0,
 0,
 2,
 2,
 0,
 2,
 4,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 2,
 0,
 0,
 0,
 6,
 2,
 0,
 0,
 0,
 0,
 2,
 2,
 2,
 0,
 0,
 0,
 2,
 1,
 2,
 2,
 1,
 0,
 0,
 0,
 3,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 3,
 2,
 0,
 0,
 0,
 1,
 0,
 2,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 3,
 0,
 0,
 4,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 2,
 1,
 2,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 2,
 1,
 1,
 1,
 3,
 2,
 1,
 0,
 2,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 2,
 0,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0]

My respond to row/column descriptions of this assignment (part 1): in the participants dataframe (e.g. name_df), it has one column labeled 'name' which stores 7 rows of entries; each row contains the name of one of the seven participants. Listed in Pandas API references, the DataFrame.shape method can be used to exhibit the dimensionality of a dataframe. 

In [38]:
print(name_df.shape)

(7, 1)


My respond to row/column descriptions of this assignment (part 2): for the messages dataframe (e.g. chess_df), it has 12 columns and 223 rows. The column of 'sender_name' stores the message senders' names respectively, the 'timestamp_ms' column contains the time points (in milliseconds)that the messages were sent. The 'content' column stores the content of each message; the 'reactions' column stores info about who reacted to sender's message.  I can't figure out the content of the 'type' column, probably a form of default entry in this case? The 'is_unsent' and 'is_taken_down' columns contains boolean variables of true/false. The 'bumped_message_metadata' contains the metadata of the sent message. The 'share', 'photo', 'gifs' and 'users' contains info about the nubmer of time the message is shared, the photos and gif animations attached to the message, etc. These four columns mostly contain empty entries. 

In [39]:
print(chess_df.shape)

(223, 12)


In [41]:
print(chess_df.columns)

Index(['sender_name', 'timestamp_ms', 'content', 'reactions', 'type',
       'is_unsent', 'is_taken_down', 'bumped_message_metadata', 'share',
       'photos', 'gifs', 'users'],
      dtype='object')


In [44]:
chess['messages'][0:1]

[{'sender_name': 'Joanna Rusch',
  'timestamp_ms': 1666374933946,
  'content': "Maybe he just wants to ride the publicity for a bit longer, even if he doesn't get any money from the lawsuit. Like, I didn't know his name before this but I certainly do now.",
  'reactions': [{'reaction': 'ð\x9f\x91\x8d', 'actor': 'Chad Larson'},
   {'reaction': 'ð\x9f\x91\x8d', 'actor': 'Chad Larson'}],
  'type': 'Generic',
  'is_unsent': False,
  'is_taken_down': False,
  'bumped_message_metadata': {'bumped_message': "Maybe he just wants to ride the publicity for a bit longer, even if he doesn't get any money from the lawsuit. Like, I didn't know his name before this but I certainly do now.",
   'is_bumped': False}}]

My respond to 'manipulate dataframe': my initial attempt is to remove/drop the 'bumped_message_metadata' column as it presents irrelevant info that woudn't be helpful to present the dataset.  The drop method listed in Pandas API references is the only way to do it. 

In [48]:
chess_df.drop(columns=['bumped_message_metadata'], inplace=True)

In [49]:
chess_df.head()

sender_name   timestamp_ms  \
0  Joanna Rusch  1666374933946   
1   Chad Larson  1666373448613   
2   Chad Larson  1666373216381   
3   Scott Pence  1666373164883   
4   Chad Larson  1666373111157   

                                             content  \
0  Maybe he just wants to ride the publicity for ...   
1  To be fair to Hans....no one wants to be assoc...   
2  He would have to prove he didn't cheat and tha...   
3  Yeah, no way. You over shoot and hope to get a...   
4  From what I see, I don't think he could win.  ...   

                                           reactions     type  is_unsent  \
0  [{'reaction': '👍', 'actor': 'Chad Larson'},...  Generic      False   
1  [{'reaction': '😆', 'actor': 'Scott Pence'},...  Generic      False   
2  [{'reaction': '👍', 'actor': 'Scott Pence'},...  Generic      False   
3  [{'reaction': '👍', 'actor': 'Chad Larson'},...  Generic      False   
4                                                     Generic      False   

   is_taken_down share photos gifs users  
0          False   NaN    NaN  NaN   NaN  
1          False   NaN    NaN  NaN   NaN  
2          False   NaN    NaN  NaN   NaN  
3          False   NaN    NaN  NaN   NaN  
4          False   NaN    NaN  NaN   NaN

My respond (cotinued): after removing the metadata column, I checked to see whether the 'share', 'photos', 'gifs' and 'users' columns contain anything other than NaN, these four columns can be removed if they contain onlh empty entries.  However, as I observed, each column does store a few non-empty entries.  Seems the content of  'photos' 'gifs' as well as the 'share' columns are originated from the message content, which appears to be similar, if not exactly the same, as text msg, twitter msg or other online chat msg.  To tidy it up, I am thinking of merging 'photos' 'gifs' into the 'content' column.  

For potential visaulizations, since the dataset looks similar to twitter data, based on my previous experience I'd propose two types of visualizations: sentiment score calculation & word clouds.  Sentence level comparative sentiment analysis can map out the mood changes throughout the conversations presented in the 'content' column and the result can be viaulized in a stacked bar chart. The word cloud visualizaiton is simple and straightforward, it hightlights the most frenquently used terms/words/expressions of the conversations and present the results in a bubble like visualization. 